### PR TITLE
Use content filtering instead of the image editor

### PR DESCRIPTION
**Pull request by jaspermdegroot**
See #83 

Update: This PR has been replaced by https://github.com/ResponsiveImagesCG/wp-tevko-responsive-images/pull/170

### CODE REVIEW

**Comment by jaspermdegroot** [comment]
I used an anonymous function in ```preg_replace_callback``` which isn't supported in PHP 5.2. That's why the tests fail. Will rewrite the code.

File: wp-tevko-responsive-images.php [hunk 1]
```diff
@@ -265,31 +265,76 @@ function tevkori_get_src_sizes( $id, $size = 'thumbnail' ) {
...
+		preg_match( '/wp-image-([0-9]+)/i', $classes[1], $id );
+		preg_match( '/size-([^\s|"]+)\s|"/i', $classes[1], $size );
+
+		$id = (int) $id[1];
+
+		// If a class with size name is present, use it.
+		if ( $size ) {
+			$size = $size[1];
+			
+		// Otherwise create an array with the values from the width and height attributes.
```
Comment (tevko) [inline wp-tevko-responsive-images.php]: What happens if there's no wp class or width / height attribute in the image html?

Comment (jaspermdegroot) [inline wp-tevko-responsive-images.php]: I would have to double check the source code but I believe the ```wp-image-*``` class and ```width``` and ```height``` attributes are always present. Of course we could implement a check and do nothing if they are not there.

Comment (tevko) [inline wp-tevko-responsive-images.php]: @jaspermdegroot there are often cases where the developer chooses to remove the WordPress classes and w/h attributes. Would we be able to retrieve the image ID if those weren't present? Or does this function fire with a higher priority, so that our calculations happen before theme customizations kick in?

Comment (jaspermdegroot) [inline wp-tevko-responsive-images.php]: > Would we be able to retrieve the image ID if those weren't present?

No, I don't think so.

> Or does this function fire with a higher priority ...

No, actually the priority is currently very low (999).

We can consider to make the priority very high instead of low, but we probably should add a check anyway to see if the image id and size could be retrieved from attributes and bail if not.

**Comment by jaspermdegroot** [comment]
I did an A/B performance test.

Server:
- Apache
- PHP version 5.6.10
- MySQL version 5.5.42

Tool:
- ApacheBench
- concurrency level 1
- 100 requests

WordPress:
- WordPress 4.3
- theme TwentyFifteen
- RICG responsive images plugin version 2.4.0
- no other plugins

WordPress installation **A** uses the plugin as it is; using the **image editor** to insert the ```srcset``` and ```sizes``` attributes in the markup.
WordPress installation **B** has the changes from this PR applied to the plugin; uses **content filtering** to insert the ```srcset``` and ```sizes``` attributes in the markup.

All images in this test have 3 sources available for the ```srcset``` (medium, large, and full).

The results, based on 3 runs:

Test post "Light" with 1271 words and **5 images**:
- A: 231 ms (230.628 ms | 231.631 ms | 231.156 ms)
- B: 245 ms (246.521 ms | 244.398 ms | 244.599 ms)

Difference: 14 ms
Average: 2.8 ms / image

Test post "Heavy" with 18 words and **25 images**:
- A: 229 ms (228.972 ms | 228.630 ms | 229.224 ms)
- B: 302 ms (302.769 ms | 302.618 ms | 301.989 ms)

Difference: 73 ms
Average: 2.9 ms / image

**Comment by jaspermdegroot** [comment]
TODO: Add code to check if ```srcset``` and ```sizes``` attributes are already present in the markup.

**Comment by jaspermdegroot** [comment]
> A few things that might improve performance is if we could grab the wp-image-{id} and size-{size} classnames in a single preg_match() instead of doing three (one for the class attribute and separate ones for each classname)

That would be nice indeed, but I would need help with that because I don't really see how we can do that.
We also have to keep in mind that the "image size class" might not be there.

> I also think we could bail early if a srcset attribute is already present.

Agreed

> If we're looking for other ways to bail early, we might be able to check to see if the file path doesn't match wp_uploads()

Good idea

> I also think we want to lower the priority value in the add_filter() function so the code executes before someone has potentially stripped off the classnames we're looking for. I'm not sure if we should go all the way to zero (lower numbers execute first) or if we should leave room for someone to override by choosing something like 5.

A priority of 5 sounds good

**Comment by joemcgill** [comment]
> That would be nice indeed, but I would need help with that because I don't really see how we can do that. We also have to keep in mind that the "image size class" might not be there.

If we did a single `preg_match` that looks for all those matches and come back with none, we can safely return early since we must have all of those pieces to execute the filter, correct?

**Comment by jaspermdegroot** [comment]
> we must have all of those pieces to execute the filter, correct?

No, we don't need the image size class because we can grab the values from the height and width attributes

**Comment by joemcgill** [comment]
Good point. So we could slim it down to 2 regex matches but that's about it.

**Comment by jaspermdegroot** [comment]
Yes, we can reduce it to 2

**Comment by tevko** [comment]
what happens if the height and width attributes are missing?

**Comment by jaspermdegroot** [comment]
> what happens if the height and width attributes are missing?

Then we bail

**Comment by jaspermdegroot** [comment]
Per discussion on Slack I changed the code to use ```attachment_url_to_postid``` to get the image ID and ```preg_match``` to get the width and height from the url.

**Comment by jaspermdegroot** [comment]
@joemcgill 

> If we're looking for other ways to bail early, we might be able to check to see if the file path doesn't match wp_uploads().

The ```attachment_url_to_postid``` core function also uses ```wp_upload_dir()``` to strip the baseurl from the path. So if we really want to optimize things we should use a customized ```attachment_url_to_postid``` function which bails if the path doesn't start with the baseurl.

Not sure if that's a good idea in regards to merging into core. What do you think?

**Comment by tevko** [comment]
@jaspermdegroot would this cause any issues for multisite installations?

**Comment by joemcgill** [comment]
@jaspermdegroot I think it would be smart to do a sanity check on the file path before passing a url to `attachment_url_to_postid()`. Honestly, I'm surprised that logic isn't present in the function itself. It may be worth filing a ticket to enhance that function in core, because right now you could pass any url you want and WP will run a query trying to find the post id that matches.

Another thing I noticed is that an edited image may have a filename that doesn't match the value of `__wp_attached_file` even taking the size pattern matching into account. Not sure if we need to handle those cases or not.

**Comment by peterwilsoncc** [comment]
Are you able to include a filter to get child attachments with the original selects on `WP_Query` or `WP_Post`. On my reading it looks like this will add a DB request for each image and avoiding this where possible would sure be dandy.

Some benchmarks would probably be in order, my instinct is it would be quicker but data would be better :).

**Comment by joemcgill** [comment]
@peterwilsoncc That's a really interesting approach. I may be mistaken, but don't attachments have to be uploaded from that particular post's edit screen in order to show up as an attachment? In other words, if someone uploaded an image directly to the media library and then inserted it into a post, would it still show up in the list of attachments for that post?

**Comment by peterwilsoncc** [comment]
@joemcgill - that's correct, it would only save requests if the author's workflow was to upload the image from the post screen. 

At present it looks like there are three SQL calls per image:

```sql
# attachment_url_to_postid
SELECT post_id FROM wp_postmeta WHERE meta_key = '_wp_attached_file' AND meta_value = '2014/07/4000.png'

# tevkori_get_srcset_array
SELECT * FROM wp_posts WHERE ID = 16 LIMIT 1
SELECT post_id, meta_key, meta_value FROM wp_postmeta WHERE post_id IN (16) ORDER BY meta_id ASC
```

Getting attachments as part of the original call should remove the second one for most images, possibly the third - I'll need to do some Nasty Hacks to check.

It could also be mitigated by updating `attachment_url_to_postid` to accept multiple URLs and batching the requests for each post.

I'm slowly coming around to the idea of dynamically adding the srcset, avoiding extra requests on each images/loop would probably convince me.

**Comment by chriscct7** [comment]
I think while its a good idea to do a performance test on 5.6, I think it'd be really prudent to also do one for 5.2. Some of the performance benefits you might see running it on 5.6, might not exist (and in fact a patch could easily be a performance detractor on 5.2). While that's unlikely, confirming that's not the case is probably a good idea.

**Comment by jaspermdegroot** [comment]
@tevko 

> would this cause any issues for multisite installations?

Is there any particular part in the code that makes you think this could cause issues with multisite installations?
Anyways, we should test it. I currently don't have multisite installation to test this on though.

**Comment by jaspermdegroot** [comment]
@joemcgill 

> I think it would be smart to do a sanity check on the file path before passing a url to attachment_url_to_postid(). Honestly, I'm surprised that logic isn't present in the function itself. It may be worth filing a ticket to enhance that function in core, because right now you could pass any url you want and WP will run a query trying to find the post id that matches.

Ok

> Another thing I noticed is that an edited image may have a filename that doesn't match the value of __wp_attached_file even taking the size pattern matching into account. Not sure if we need to handle those cases or not.

I think we should try to handle those cases, but it doesn't feel rock solid to simply strip any "-e" followed by 13 digits from the end of the file name since it's perfectly valid to upload an image with that pattern in the filename. Probably an edge case but still.

**Comment by jaspermdegroot** [comment]
The extra SQL request was also my concern when we discussed using the url to get the ID instead of grabbing it from the class name and tests do show that is a bit slower, but I agree with @tevko and @joemcgill that it is more reliable.

If we can avoid the extra request that would be great. Thanks for your input @peterwilsoncc . I will look into your approach.

**Comment by jaspermdegroot** [comment]
@chriscct7 

Good point. I will test performance against PHP 5.2 as well.

**Comment by peterwilsoncc** [comment]
I've done some work going directly from the image URLs to getting the attachment objects. It has less DB calls but whether it is more efficient overall will need further benchmarking. 

I've been working on the branch [peterwilsoncc:content-filtering-pwcc](https://github.com/peterwilsoncc/wp-tevko-responsive-images/tree/content-filtering-pwcc) [[diff with jaspermdegroot:content-filtering](https://github.com/jaspermdegroot/wp-tevko-responsive-images/compare/content-filtering...peterwilsoncc:content-filtering-pwcc)]

Callback is now in a class which requires a separate instance for each post, there's some shuffling around to make sure all image sizes return an ID, all of this leads me to question how efficient it is.

All in all, not sure how much help this has been for y'all.

**Comment by jaspermdegroot** [comment]
This PR is replaced by https://github.com/ResponsiveImagesCG/wp-tevko-responsive-images/pull/170
